server/proof_fetcher_test.go

```diff
@@ -0,0 +1,360 @@
+	storageNodes []storage.Node     // nodes returned by storage layer
+	storageError error              // error returned by storage layer
+	result       []storage.Node     // expected result with dupes reinstated
+}
+
+// Contents of these don't really matter as long as they have distinct IDs
+var f11 = merkle.NodeFetch{NodeID: mustCreateNodeID(1, 1)}
+var f12 = merkle.NodeFetch{NodeID: mustCreateNodeID(1, 2)}
+var f31 = merkle.NodeFetch{NodeID: mustCreateNodeID(3, 1)}
+
+var n11 = storage.Node{NodeID: f11.NodeID, NodeRevision: 37}
+var n12 = storage.Node{NodeID: f12.NodeID, NodeRevision: 37}
+var n31 = storage.Node{NodeID: f31.NodeID, NodeRevision: 37}
+
+var dedupTests = []dedupTest{
+	{
+		desc:         "no dupes",
+		input:        []merkle.NodeFetch{f11, f12, f31},
+		storageIDs:   []storage.NodeID{f11.NodeID, f12.NodeID, f31.NodeID},
+		storageNodes: []storage.Node{n11, n12, n31},
+		result:       []storage.Node{n11, n12, n31},
+	},
+	{
+		desc:         "one dupe",
+		input:        []merkle.NodeFetch{f11, f12, f31, f11},
+		storageIDs:   []storage.NodeID{f11.NodeID, f12.NodeID, f31.NodeID},
+		storageNodes: []storage.Node{n11, n12, n31},
+		result:       []storage.Node{n11, n12, n31, n11},
+	},
+	{
+		desc:         "multi dupe",
+		input:        []merkle.NodeFetch{f11, f12, f11, f11, f31, f11},
+		storageIDs:   []storage.NodeID{f11.NodeID, f12.NodeID, f31.NodeID},
+		storageNodes: []storage.Node{n11, n12, n31},
+		result:       []storage.Node{n11, n12, n11, n11, n31, n11},
+	},
+	{
+		desc:         "storage fail",
+		input:        []merkle.NodeFetch{f11, f12, f11, f11, f31, f11},
+		storageIDs:   []storage.NodeID{f11.NodeID, f12.NodeID, f31.NodeID},
+		storageNodes: []storage.Node{},
+		storageError: errors.New("storage"),
+	},
+	{
+		desc:         "storage wrong node",
+		input:        []merkle.NodeFetch{f11, f12, f11, f11, f31, f11},
+		storageIDs:   []storage.NodeID{f11.NodeID, f12.NodeID, f31.NodeID},
+		storageNodes: []storage.Node{},
+		storageError: errors.New("storage"),
+	},
+}
+
+func TestRehasher(t *testing.T) {
+	for _, rehashTest := range rehashTests {
+		r := newRehasher()
+		for i, node := range rehashTest.nodes {
+			r.process(node, rehashTest.fetches[i])
+		}
+
+		want := rehashTest.output
+		got, err := r.rehashedProof(rehashTest.index)
+
+		if err != nil {
+			t.Fatalf("rehash test %s unexpected error: %v", rehashTest.desc, err)
+		}
+
+		if !proto.Equal(&got, &want) {
+			t.Errorf("rehash test %s:\ngot: %v\nwant: %v", rehashTest.desc, got, want)
+		}
+	}
+}
+
+func TestDedupFetcher(t *testing.T) {
+	for _, dedupTest := range dedupTests {
+		ctrl := gomock.NewController(t)
+		tx := storage.NewMockLogTX(ctrl)
+		tx.EXPECT().GetMerkleNodes(int64(37), dedupTest.storageIDs).Return(dedupTest.storageNodes, dedupTest.storageError)
+		nodes, err := dedupAndFetchNodes(tx, 37, dedupTest.input)
+
+		if err == nil && dedupTest.storageError != nil {
+			t.Fatalf("%s: got nil, want error: %v", dedupTest.desc, err)
```

Comment:
Wasn't sure because it's just going to dump hex strings that don't match. I think printing a bunch of these is no extra help but willing to be convinced otherwise.